test/integration/dynamic-reducers.spec.js

```diff
@@ -0,0 +1,165 @@
+                  // Only patching if storeState exists makes this test
+                  // work with multiple React-Redux approaches
+                  if (contextState && !(key in contextState)) {
+                    shouldPatch = true
+                  }
+                }
+
+                if (shouldInject) {
+                  injectReducers(reducers)
+                }
+
+                if (shouldPatch) {
+                  // A safer way to do this would be to patch the storeState
+                  // manually with the state from the new reducers, since
+                  // this would better avoid tearing in a future concurrent world
+                  const patchedReduxContext = {
+                    ...reduxContext,
+                    storeState: reduxContext.store.getState()
+                  }
+                  return (
+                    <ReactReduxContext.Provider value={patchedReduxContext}>
+                      {children}
+                    </ReactReduxContext.Provider>
+                  )
+                }
+
+                return children
+              }}
+            </ReactReduxContext.Consumer>
+          )}
+        </InjectReducersContext.Consumer>
+      )
+    }
+
+    const initialReducer = {
+      initial: (state = { greeting: 'Hello world' }) => state
+    }
+    const dynamicReducer = {
+      dynamic: (state = { greeting: 'Hello dynamic world' }) => state
+    }
+
+    function Greeter({ greeting }) {
+      return <div>{greeting}</div>
+    }
+
+    const InitialGreeting = connect(state => ({
+      greeting: state.initial.greeting
+    }))(Greeter)
+    const DynamicGreeting = connect(state => ({
+      greeting: state.dynamic.greeting
+    }))(Greeter)
+
+    function createInjectReducers(store, initialReducer) {
+      let reducers = initialReducer
+      return function injectReducers(newReducers) {
+        reducers = { ...reducers, ...newReducers }
+        store.replaceReducer(combineReducers(reducers))
+      }
+    }
+
+    let store
+    let injectReducers
+
+    beforeEach(() => {
+      // These could be singletons on the client, but
+      // need to be separate per request on the server
+      store = createStore(combineReducers(initialReducer))
+      injectReducers = createInjectReducers(store, initialReducer)
+    })
+
+    it('should render child with initial state on the client', () => {
+      const { getByText } = rtl.render(
+        <Provider store={store}>
+          <InjectReducersContext.Provider value={injectReducers}>
+            <InitialGreeting />
+            <ExtraReducersProvider reducers={dynamicReducer}>
+              <DynamicGreeting />
+            </ExtraReducersProvider>
+          </InjectReducersContext.Provider>
+        </Provider>
+      )
+
+      getByText('Hello world')
+      getByText('Hello dynamic world')
+    })
+    it('should render child with initial state on the server', () => {
```

Comment:
Sadly the test-environment can only differ between files, not single tests. If you think it makes more sense, I'll refactor the implementation into a util-file and move this test over to the SSR-suite where the true node test-environment is running, or a new `dynamic-reducers.server.spec.js`.